server/src/main/java/org/elasticsearch/index/mapper/TypeFieldMapper.java

```diff
@@ -174,6 +179,36 @@ public Query termsQuery(List<?> values, QueryShardContext context) {
             }
         }
 
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (hasDocValues()) {
+                return new TermRangeQuery(name(), lowerTerm == null ? null : indexedValueForSearch(lowerTerm),
+                        upperTerm == null ? null : indexedValueForSearch(upperTerm), includeLower, includeUpper);
```

Comment:
@jpountz could you take a look at this backport also, please? This was one of my open questions in #31756, whether we need this for cases where we have old 5.x indices with multiple types.

Reply:
we need to take care of 5.x indices indeed, but the condition in the if statement should be `context.getIndexSettings().isSingleType() == false`